pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java

```diff
@@ -0,0 +1,258 @@
+                    packet = handleFragmentation(ip);
+                    if (packet == null) {
+                        return true;
+                    }
+                }
+            }
+
+            if (packet.hasProtocol(Protocol.TCP) &&
+                    (packet.hasProtocol(Protocol.IPv4) || packet.hasProtocol(Protocol.IPv6))) {
+                this.processFrame(packet);
+            }
+
+        } catch (final IOException | PacketParseException e) {
+            e.printStackTrace();
+        }
+
+        return true;
+    }
+
+    public void processFrame(final Packet frame) throws PacketParseException {
+        try {
+            final IPPacket ipPacket =
+                    frame.hasProtocol(Protocol.IPv4) ?
+                            (IPv4Packet) frame.getPacket(Protocol.IPv4)
+                            : (frame.hasProtocol(Protocol.IPv6) ? (IPv6Packet) frame.getPacket(Protocol.IPv6) : null);
+
+            final TCPPacket tcpPacket = (TCPPacket) frame.getPacket(Protocol.TCP);
+
+            if (ipPacket == null || tcpPacket == null){
+                throw new NullPointerException("tcp or ip packet was null when processed");
+            }
+
+            final TransportStreamId pktStreamId = new TransportStreamId(tcpPacket);
+
+            TcpStream stream = activeTcpStreams.get(pktStreamId);
+            stream = (stream == null) ? activeTcpStreams.get(pktStreamId.oppositeFlowDirection()) : stream;
+
+            if (stream == null) {
+                startNewStream(tcpPacket);
+            } else {
+                final boolean isAlreadyClosed = stream.Ended();
+                stream.addPacket(tcpPacket);
+                this.notifyPacketReceived(stream, tcpPacket);
+                if (!isAlreadyClosed && stream.Ended()){ // call endStream only once when the last packet closed it
+                    this.notifyEndStream(stream);
+                }
+            }
+
+        } catch (Exception e){
+            e.printStackTrace();
+        }
+    }
+
+    private IPPacket handleFragmentation(final IPPacket ipPacket) {
+        if (this.fragmentListener == null) {
+            return null;
+        }
+        try {
+            return this.fragmentListener.handleFragment(ipPacket);
+        } catch (final Throwable t) {
+            logger.warn("Exception thrown by FragmentListener when processing the IP frame", t);
+        }
+        return null;
+    }
+
+    private void notifyStartStream(final TcpStream stream, final TCPPacket pkt) {
+        if (this.tcpListener != null) {
+            this.tcpListener.startStream(stream, pkt);
+        }
+    }
+
+    private void notifyPacketReceived(final TcpStream stream, final TCPPacket pkt) {
+        if (this.tcpListener != null) {
+            this.tcpListener.packetReceived(stream, pkt);
+        }
+    }
+
+    private void notifyEndStream(final TcpStream stream) {
+        if (this.tcpListener != null) {
+            this.tcpListener.endStream(stream);
+        }
```

Comment:
a technique I often deploy to avoid having to do null-checks all the time is just to have a dummy, in this case, TcpListener that'll do nothing. That way, I do not have to check for null as my dummy would just be a NOOP...